File: MEM/models/reinforcement_learning/rl_state.py
"""
Reinforcement Learning State
=============================
State representation for RL agent.

Mirrors C# implementation in:
backend/AlgoTrendy.TradingEngine/Models/ReinforcementLearning/RLState.cs

Contains:
- 22 features from ASFeatures
- Inventory state
- Episode context (time remaining, step number)
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Dict, Any
import numpy as np
import json

from ..market_making import ASFeatures, InventoryState


@dataclass
class RLState:
    """
    Reinforcement Learning state representation.
    Contains all features and context needed by the RL agent to make decisions.
    """

    symbol: str
    """Symbol being traded"""

    timestamp: datetime
    """Timestamp of this state"""

    features: ASFeatures
    """Features for RL agent (22 features)"""

    inventory: InventoryState
    """Current inventory state"""

    current_price: float
    """Current market price (mid price from order book)"""

    time_remaining: float = 1.0
    """Time remaining in trading session (0.0 to 1.0). 1.0=start, 0.0=end"""

    step_number: int = 0
    """Episode step number (increments each action)"""

    is_terminal: bool = False
    """Is this a terminal state (end of episode)"""

    metadata: Optional[Dict[str, Any]] = None
    """Additional metadata (optional)"""

    # Constants
    STATE_DIMENSION: int = field(default=22, init=False, repr=False)
    """State dimensionality (should always be 22)"""

    def to_array(self) -> np.ndarray:
        """
        Converts state to array for RL agent input.
        Uses ASFeatures.to_array() - returns 22 features.

        Returns:
            Array of 22 state features
        """
        return self.features.to_array()

    def to_extended_array(self) -> np.ndarray:
        """
        Gets extended state array including additional context.
        22 features + 3 context = 25 dimensions.

        Returns:
            Extended state array (25 dimensions)
        """
        features = self.features.to_array()

        # Add context (3 additional dimensions)
        context = np.array([
            self.time_remaining,
            self.inventory.risk_level / 100.0,  # Normalized 0-1
            float(self.step_number)
        ], dtype=np.float32)

        return np.concatenate([features, context])

    @classmethod
    def create(
        cls,
        symbol: str,
        features: ASFeatures,
        inventory: InventoryState,
        current_price: float,
        time_remaining: float = 1.0,
        step_number: int = 0
    ) -> 'RLState':
        """
        Creates RLState from ASFeatures and InventoryState.

        Args:
            symbol: Trading symbol
            features: AS features
            inventory: Inventory state
            current_price: Current market price
            time_remaining: Time remaining in session (default: 1.0)
            step_number: Episode step number (default: 0)

        Returns:
            New RLState instance
        """
        return cls(
            symbol=symbol,
            timestamp=datetime.utcnow(),
            features=features,
            inventory=inventory,
            current_price=current_price,
            time_remaining=time_remaining,
            step_number=step_number,
            is_terminal=False
        )

    @classmethod
    def create_terminal(cls, current_state: 'RLState') -> 'RLState':
        """
        Creates a terminal state (end of episode).

        Args:
            current_state: Current state to mark as terminal

        Returns:
            Terminal state
        """
        return cls(
            symbol=current_state.symbol,
            timestamp=datetime.utcnow(),
            features=current_state.features,
            inventory=current_state.inventory,
            current_price=current_state.current_price,
            time_remaining=0.0,
            step_number=current_state.step_number,
            is_terminal=True,
            metadata=current_state.metadata
        )

    def to_dict(self) -> dict:
        """
        Convert to dictionary for JSON serialization.

        Returns:
            Dictionary representation
        """
        return {
            "symbol": self.symbol,
            "timestamp": self.timestamp.isoformat(),
            "features": self.features.to_dict(),
            "inventory": self.inventory.to_dict(),
            "current_price": self.current_price,
            "time_remaining": self.time_remaining,
            "step_number": self.step_number,
            "is_terminal": self.is_terminal,
            "metadata": self.metadata
        }

    def to_json(self) -> str:
        """
        Convert to JSON string.

        Returns:
            JSON string representation
        """
        return json.dumps(self.to_dict(), indent=2)

    @classmethod
    def from_dict(cls, data: dict) -> 'RLState':
        """
        Create RLState from dictionary.

        Args:
            data: Dictionary with state data

        Returns:
            RLState instance
        """
        # Parse timestamp
        timestamp = data['timestamp']
        if isinstance(timestamp, str):
            timestamp = datetime.fromisoformat(timestamp.replace('Z', '+00:00'))

        return cls(
            symbol=data['symbol'],
            timestamp=timestamp,
            features=ASFeatures.from_dict(data['features']),
            inventory=InventoryState.from_dict(data['inventory']),
            current_price=data['current_price'],
            time_remaining=data.get('time_remaining', 1.0),
            step_number=data.get('step_number', 0),
            is_terminal=data.get('is_terminal', False),
            metadata=data.get('metadata')
        )

    @classmethod
    def from_json(cls, json_str: str) -> 'RLState':
        """
        Create RLState from JSON string.

        Args:
            json_str: JSON string

        Returns:
            RLState instance
        """
        data = json.loads(json_str)
        return cls.from_dict(data)

    def clone(self) -> 'RLState':
        """
        Deep clone.

        Returns:
            Copy of RLState
        """
        return RLState(
            symbol=self.symbol,
            timestamp=self.timestamp,
            features=self.features,  # ASFeatures is immutable (dataclass)
            inventory=self.inventory.clone(),
            current_price=self.current_price,
            time_remaining=self.time_remaining,
            step_number=self.step_number,
            is_terminal=self.is_terminal,
            metadata=dict(self.metadata) if self.metadata else None
        )

    def __str__(self) -> str:
        terminal_str = " [TERMINAL]" if self.is_terminal else ""
        return (
            f"RLState[{self.symbol}] Step={self.step_number}, "
            f"Price={self.current_price:.2f}, Inv={self.inventory.inventory_percent:.0%}, "
            f"Risk={self.inventory.risk_category}, T={self.time_remaining:.0%}{terminal_str}"
        )

    def __repr__(self) -> str:
        return (
            f"RLState(symbol='{self.symbol}', step={self.step_number}, "
            f"price={self.current_price:.2f}, terminal={self.is_terminal})"
        )

File: backend/AlgoTrendy.TradingEngine/Models/ReinforcementLearning/RLState.cs
using AlgoTrendy.TradingEngine.Models.MarketMaking;

namespace AlgoTrendy.TradingEngine.Models.ReinforcementLearning;

/// <summary>
/// Reinforcement Learning state representation
/// Contains all features and context needed by the RL agent to make decisions
/// </summary>
public class RLState
{
    /// <summary>
    /// Symbol being traded
    /// </summary>
    public required string Symbol { get; init; }

    /// <summary>
    /// Timestamp of this state
    /// </summary>
    public required DateTime Timestamp { get; init; }

    /// <summary>
    /// Features for RL agent (22 features)
    /// </summary>
    public required ASFeatures Features { get; init; }

    /// <summary>
    /// Current inventory state
    /// </summary>
    public required InventoryState Inventory { get; init; }

    /// <summary>
    /// Current market price (mid price from order book)
    /// </summary>
    public decimal CurrentPrice { get; init; }

    /// <summary>
    /// Time remaining in trading session (0.0 to 1.0)
    /// 1.0 = session start, 0.0 = session end
    /// </summary>
    public decimal TimeRemaining { get; init; } = 1.0m;

    /// <summary>
    /// Episode step number (increments each action)
    /// </summary>
    public int StepNumber { get; init; } = 0;

    /// <summary>
    /// Is this a terminal state (end of episode)
    /// </summary>
    public bool IsTerminal { get; init; } = false;

    /// <summary>
    /// Additional metadata (optional)
    /// </summary>
    public Dictionary<string, object>? Metadata { get; init; }

    /// <summary>
    /// State dimensionality (should always be 22)
    /// </summary>
    public const int StateDimension = 22;

    /// <summary>
    /// Converts state to array for RL agent input
    /// Uses ASFeatures.ToArray() - returns 22 features
    /// </summary>
    /// <returns>Array of 22 state features</returns>
    public double[] ToArray()
    {
        return Features.ToArray();
    }

    /// <summary>
    /// Gets extended state array including additional context
    /// 22 features + 3 context = 25 dimensions
    /// </summary>
    /// <returns>Extended state array (25 dimensions)</returns>
    public double[] ToExtendedArray()
    {
        var features = Features.ToArray();
        var extended = new double[25];

        // Copy 22 features
        Array.Copy(features, extended, 22);

        // Add context (3 additional dimensions)
        extended[22] = (double)TimeRemaining;
        extended[23] = (double)Inventory.RiskLevel / 100.0; // Normalized 0-1
        extended[24] = (double)StepNumber;

        return extended;
    }

    /// <summary>
    /// Creates RLState from ASFeatures and InventoryState
    /// </summary>
    /// <param name="symbol">Trading symbol</param>
    /// <param name="features">AS features</param>
    /// <param name="inventory">Inventory state</param>
    /// <param name="currentPrice">Current market price</param>
    /// <param name="timeRemaining">Time remaining in session</param>
    /// <param name="stepNumber">Episode step number</param>
    /// <returns>New RLState instance</returns>
    public static RLState Create(
        string symbol,
        ASFeatures features,
        InventoryState inventory,
        decimal currentPrice,
        decimal timeRemaining = 1.0m,
        int stepNumber = 0)
    {
        return new RLState
        {
            Symbol = symbol,
            Timestamp = DateTime.UtcNow,
            Features = features,
            Inventory = inventory,
            CurrentPrice = currentPrice,
            TimeRemaining = timeRemaining,
            StepNumber = stepNumber,
            IsTerminal = false
        };
    }

    /// <summary>
    /// Creates a terminal state (end of episode)
    /// </summary>
    /// <param name="currentState">Current state to mark as terminal</param>
    /// <returns>Terminal state</returns>
    public static RLState CreateTerminal(RLState currentState)
    {
        return new RLState
        {
            Symbol = currentState.Symbol,
            Timestamp = DateTime.UtcNow,
            Features = currentState.Features,
            Inventory = currentState.Inventory,
            CurrentPrice = currentState.CurrentPrice,
            TimeRemaining = 0.0m,
            StepNumber = currentState.StepNumber,
            IsTerminal = true,
            Metadata = currentState.Metadata
        };
    }

    /// <summary>
    /// Deep clone
    /// </summary>
    public RLState Clone()
    {
        return new RLState
        {
            Symbol = Symbol,
            Timestamp = Timestamp,
            Features = Features,
            Inventory = Inventory.Clone(),
            CurrentPrice = CurrentPrice,
            TimeRemaining = TimeRemaining,
            StepNumber = StepNumber,
            IsTerminal = IsTerminal,
            Metadata = Metadata != null ? new Dictionary<string, object>(Metadata) : null
        };
    }

    public override string ToString()
    {
        var terminalStr = IsTerminal ? " [TERMINAL]" : "";
        return $"RLState[{Symbol}] Step={StepNumber}, " +
               $"Price={CurrentPrice:F2}, Inv={Inventory.InventoryPercent:P0}, " +
               $"Risk={Inventory.RiskCategory}, T={TimeRemaining:P0}{terminalStr}";
    }
}
